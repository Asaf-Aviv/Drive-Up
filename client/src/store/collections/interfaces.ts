import { MovieShowcase } from '../movies/interfaces';

export interface Collection {
  id: number;
  name: string;
  overview: string;
  poster_path: string | null;
  backdrop_path: string | null;
  parts: MovieShowcase[];
}
