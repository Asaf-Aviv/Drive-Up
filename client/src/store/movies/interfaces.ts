export interface Results<T> {
  results: T;
  page: number;
  total_pages: number;
  total_results: number;
}

export interface Genre {
  id: number;
  name: string;
}

export interface Series {
  id: number;
  name: string;
  poster_path: string;
  backdrop_path: string;
}

export interface Company {
  name: string;
  id: number;
  logo_path: string | null;
  origin_country: string;
}

export interface Image {
  file_path: string;
  width: number;
  height: number;
}

export interface Video {
  id: string;
  key: string;
  name: string;
  site: string;
  type: string;
}

export interface MovieShowcase {
  poster_path: string | null;
  adult: boolean;
  overview: string;
  release_date: string;
  genre_ids: number[];
  id: number;
  original_title: string;
  original_language: string;
  title: string;
  backdrop_path: string | null;
  popularity: number;
  vote_count: number;
  video: boolean;
  vote_average: number;
}

export interface Movie {
  adult: boolean;
  backdrop_path: string | null;
  belongs_to_collection: Series | null;
  budget: number;
  genres: Genre[];
  homepage: string | null;
  id: number;
  images: {
    posters: Image[];
    backdrops: Image[];
  };
  videos: {
    results: Video[];
  };
  similar: Results<MovieShowcase[]>;
  recommendations: Results<MovieShowcase[]>;
  original_language: string;
  original_title: string;
  overview: string;
  popularity: number;
  poster_path: string | null;
  production_companies: Company[];
  release_date: string;
  revenue: number;
  runtime: number | null;
  tagline: string | null;
  title: string;
  video: boolean;
  vote_average: number;
  vote_count: number;
}
