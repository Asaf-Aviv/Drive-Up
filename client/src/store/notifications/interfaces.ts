type NotificationVariant = 'error' | 'success' | 'info' | 'warning';

export interface Notification {
  id: string;
  variant: NotificationVariant;
  message: string;
}
