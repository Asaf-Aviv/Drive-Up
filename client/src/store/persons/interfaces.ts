export interface PersonSummary {
  profile_path: string;
  adult: boolean;
  id: number;
  name: string;
  popularity: number;
  known_for_department: string;
  gender: 0 | 1 | 2;
}
