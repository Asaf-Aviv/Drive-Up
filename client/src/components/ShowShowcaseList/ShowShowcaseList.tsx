import React, { useEffect, useRef } from 'react';
import { useDispatch } from 'react-redux';
import { List, ListItem } from '@material-ui/core';
import useShallowEqualSelector from '../../hooks/useShallowEqualSelector';
import { useOnVisibilityTrigger } from '../../hooks/useOnVisibilityTrigger';
import { ShowShowcase as IShowShowcase } from '../../store/shows/interfaces';
import RetryButton from '../RetryButton';
import { requestShows } from '../../store/shows/actions';
import ShowShowcase from '../ShowShowcase';

const ShowShowcaseList: React.FC = () => {
  const {
    page, loading, shows, error,
  } = useShallowEqualSelector(state => state.shows);
  const dispatch = useDispatch();
  const fetchNextPageTrigger = useRef<HTMLElement>(null);

  const fetchNextShowsPage = () => {
    if (shows.length) return;

    dispatch(requestShows(page + 1));
  };

  useEffect(fetchNextShowsPage, []);

  useOnVisibilityTrigger(fetchNextPageTrigger, fetchNextShowsPage);

  const renderShowsShowcase = (show: IShowShowcase) => (
    <ListItem key={show.id}>
      <ShowShowcase {...show} />
    </ListItem>
  );

  return (
    <main>
      <List>
        {shows.map(renderShowsShowcase)}
        {!error && !loading && <span ref={fetchNextPageTrigger} />}
        {loading && <span>Loading...</span>}
        {error && <RetryButton onClick={fetchNextShowsPage} />}
      </List>
    </main>
  );
};

export default ShowShowcaseList;
