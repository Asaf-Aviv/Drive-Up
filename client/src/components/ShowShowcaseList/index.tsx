import ShowShowcaseList from './ShowShowcaseList';

export default ShowShowcaseList;
