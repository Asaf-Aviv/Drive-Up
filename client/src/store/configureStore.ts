import {
  createStore,
  applyMiddleware,
  Middleware,
  Reducer,
} from 'redux';
import thunkMiddleware from 'redux-thunk';
import logger from 'redux-logger';
import createSagaMiddleware from 'redux-saga';
import { composeWithDevTools } from 'redux-devtools-extension';
import rootSaga from './sagas/rootSaga';

export default function configureStore(rootReducer: Reducer, preloadedState = {}) {
  const devMode = process.env.NODE_ENV !== 'production';

  const sagaMiddleware = createSagaMiddleware();

  const middlewares: Middleware[] = [sagaMiddleware, thunkMiddleware];

  if (devMode) {
    middlewares.push(logger);
  }

  const middleWareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(
    rootReducer,
    preloadedState,
    devMode
      ? composeWithDevTools(middleWareEnhancer)
      : middleWareEnhancer,
  );

  sagaMiddleware.run(rootSaga);

  return store;
}
