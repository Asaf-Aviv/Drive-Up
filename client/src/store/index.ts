import {
  combineReducers,
  createStore,
  applyMiddleware,
  Middleware,
} from 'redux';
import thunkMiddleware from 'redux-thunk';
import logger from 'redux-logger';
import createSagaMiddleware, { SagaMiddleware } from 'redux-saga';
import { composeWithDevTools } from 'redux-devtools-extension';
import notificationsReducer from './notifications/reducers';
import moviesReducer from './movies/reducers';

const rootReducer = combineReducers({
  notifications: notificationsReducer,
  movies: moviesReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

export default () => {
  const devMode = process.env.NODE_ENV !== 'production';

  const sagaMiddleware = createSagaMiddleware();

  const middlewares: Middleware[] = [sagaMiddleware, thunkMiddleware];

  if (devMode) {
    middlewares.push(logger);
  }

  const middleWareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(
    rootReducer,
    devMode
      ? composeWithDevTools(middleWareEnhancer)
      : middleWareEnhancer,
  );

  // sagaMiddleware.run()

  return store;
};
