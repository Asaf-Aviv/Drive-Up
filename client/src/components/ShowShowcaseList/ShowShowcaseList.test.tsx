import React from 'react';
import 'intersection-observer';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { Store } from 'redux';
import { MemoryRouter } from 'react-router';
import ShowShowcaseList from './ShowShowcaseList';
import configureStore from '../../store/configureStore';
import { rootReducer } from '../../store';
import { ShowShowcase } from '../../store/shows/interfaces';
import showsReducer from '../../store/shows/reducers';

const shows: ShowShowcase[] = [
  {
    original_name: 'American Dad!',
    name: 'American Dad!',
    popularity: 92.504,
    genre_ids: [],
    vote_count: 734,
    first_air_date: '2005-02-06',
    backdrop_path: '/vddYkC55H3DqWK5cG5hepSnhzGP.jpg',
    original_language: 'en',
    origin_country: [],
    id: 1433,
    vote_average: 6.1,
    overview: 'The series focuses on an eccentric motley',
    poster_path: '/eo2Xu4UWXHE8UlBlAktNiSsAmfx.jpg',
  },
  {
    original_name: 'Gotham',
    name: 'Gotham',
    popularity: 113.872,
    genre_ids: [],
    vote_count: 1157,
    first_air_date: '2014-09-22',
    backdrop_path: '/mKBP1OCgCG0jw8DwVYlnYqVILtc.jpg',
    original_language: 'en',
    origin_country: [],
    id: 60708,
    vote_average: 6.9,
    overview: 'Before there was Batman, there was',
    poster_path: '/4XddcRDtnNjYmLRMYpbrhFxsbuq.jpg',
  },
];

describe('ShowsShowcaseList component', () => {
  let store: Store;

  it('renders correctly', () => {
    store = configureStore(rootReducer);

    const wrapper = mount(
      <Provider store={store}>
        <ShowShowcaseList />
      </Provider>
    );

    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('should render Loading... when loading state is true', () => {
    store = configureStore(showsReducer, {
      shows: {
        page: 1,
        loading: true,
        error: false,
        shows: [],
      },
    });

    const wrapper = mount(
      <Provider store={store}>
        <ShowShowcaseList />
      </Provider>
    );

    const spanText = wrapper.find('span').text();

    expect(spanText).toEqual('Loading...');
  });

  it('should render RetryButton component when error state is true', () => {
    store = configureStore(showsReducer, {
      shows: {
        page: 1,
        loading: false,
        error: true,
        shows: [],
      },
    });

    const wrapper = mount(
      <Provider store={store}>
        <ShowShowcaseList />
      </Provider>
    );

    expect(wrapper.find('RetryButton')).toHaveLength(1);
  });

  it('should render shows list', () => {
    store = configureStore(showsReducer, {
      shows: {
        page: 1,
        loading: false,
        error: false,
        shows,
      },
    });

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <ShowShowcaseList />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('ShowShowcase')).toHaveLength(2);
  });
});
