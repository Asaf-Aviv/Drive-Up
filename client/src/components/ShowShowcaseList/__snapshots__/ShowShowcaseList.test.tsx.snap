// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowsShowcaseList component renders correctly 1`] = `
"<Provider store={{...}}>
  <ShowShowcaseList>
    <main>
      <WithStyles(ForwardRef(List))>
        <ForwardRef(List) classes={{...}}>
          <ul className=\\"MuiList-root MuiList-padding\\">
            <span>
              Loading...
            </span>
          </ul>
        </ForwardRef(List)>
      </WithStyles(ForwardRef(List))>
    </main>
  </ShowShowcaseList>
</Provider>"
`;
